Mathlib/Analysis/Convex/EGauge.lean

```diff
@@ -173,22 +196,118 @@ lemma egauge_smul_right (h : c = 0 → s.Nonempty) (x : E) :
+        simp [hc0 i hi]
+  refine egauge_lt_iff.2 ⟨c₀, ?_, hc₀r⟩
+  suffices x ∈ I.pi fun i ↦ c₀ • U i by
+    rcases hc₀ with hc₀ | rfl
+    · simpa only [← image_smul, ← piMap_image_pi (fun _ _ ↦ MulAction.surjective₀ hc₀)] using this
+    · simpa only [← image_smul, ← piMap_image_univ_pi] using this
+  intro i hi
+  exact (hU i hi).smul_mono (hc₀I i hi) (hc i hi)
+
+/-- The extended gauge of a point `x` in an indexed product
+with respect to a product of finitely many balanced sets `U i`, `i ∈ I`,
+(and the whole spaces for the other indices)
+is the supremum of the extended gauges of the components of `x`
+with respect to the corresponding balanced set.
+
+This version assumes the following technical condition:
+- either `I` is the universal set;
+- or one of `x i`, `i ∈ I`, is nonzero;
+- or `𝕜` is nontrivially normed.
+-/
+theorem egauge_univ_pi [Finite ι] {U : ∀ i, Set (E i)} (hU : ∀ i, Balanced 𝕜 (U i)) (x : ∀ i, E i) :
+    egauge 𝕜 (univ.pi U) x = ⨆ i, egauge 𝕜 (U i) (x i) :=
+  egauge_pi' finite_univ (fun i _ ↦ hU i) x (.inl rfl) |>.trans <| by simp
+
+end Pi
+
+variable {𝕜 : Type*} [NormedDivisionRing 𝕜] {E : Type*} [AddCommGroup E] [Module 𝕜 E]
+  {F : Type*} [AddCommGroup F] [Module 𝕜 F]
+
+/-- The extended gauge of a point `(a, b)` with respect to the product of balanced sets `U` and `V`
+is equal to the maximum of the extended gauges of `a` with respect to `U`
+and `b` with respect to `V`.
+-/
+theorem egauge_prod_mk {U : Set E} {V : Set F}
+    (hU : Balanced 𝕜 U) (hV : Balanced 𝕜 V) (a : E) (b : F) :
+    egauge 𝕜 (U ×ˢ V) (a, b) = max (egauge 𝕜 U a) (egauge 𝕜 V b) := by
+  refine le_antisymm ?_ (max_le (le_egauge_prod_left _ _ _ _) (le_egauge_prod_right _ _ _ _))
+  refine le_of_forall_lt' fun r hr ↦ ?_
+  simp only [max_lt_iff, egauge_lt_iff, smul_set_prod, mk_mem_prod] at hr ⊢
+  rcases hr with ⟨⟨x, hx, hxr⟩, ⟨y, hy, hyr⟩⟩
+  cases le_total ‖x‖ ‖y‖ with
+  | inl hle => exact ⟨y, ⟨hU.smul_mono hle hx, hy⟩, hyr⟩
+  | inr hle => exact ⟨x, ⟨hx, hV.smul_mono hle hy⟩, hxr⟩
 
 theorem egauge_add_add_le {U V : Set E} (hU : Balanced 𝕜 U) (hV : Balanced 𝕜 V) (a b : E) :
     egauge 𝕜 (U + V) (a + b) ≤ max (egauge 𝕜 U a) (egauge 𝕜 V b) := by
-  refine le_of_forall_lt' fun c hc ↦ ?_
-  simp only [max_lt_iff, egauge_lt_iff] at hc ⊢
-  rcases hc with ⟨⟨x, hx, hxc⟩, ⟨y, hy, hyc⟩⟩
-  wlog hxy : ‖x‖ ≤ ‖y‖ generalizing a b x y U V
-  · simpa only [add_comm] using this hV hU b a y hy hyc x hx hxc (le_of_not_le hxy)
-  refine ⟨y, ?_, hyc⟩
-  rw [smul_add]
-  exact add_mem_add (hU.smul_mono hxy hx) hy
-
-end VectorSpace
+  rw [← egauge_prod_mk hU hV a b, ← add_image_prod]
+  exact MapsTo.egauge_le 𝕜 (LinearMap.fst 𝕜 E E + LinearMap.snd 𝕜 E E) (mapsTo_image _ _) (a, b)
+
+end NormedDivisionRing
+
+section NontriviallyNormedField
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  {ι : Type*} {E : ι → Type*} [∀ i, AddCommGroup (E i)] [∀ i, Module 𝕜 (E i)]
+
+theorem egauge_pi {I : Set ι} {U : ∀ i, Set (E i)}
```

Comment:
Could yo uadd an `egauge_pi_univ` shorthand in terms of `Finite`, for convenience?

Reply:
There is `egauge_univ_pi` above (I should fix its doc string).

Reply:
Oh, I expected them to be in the other order.

Reply:
`egauge_univ_pi` doesn't assume that `K` is a nontrivially normed field, just a normed division ring.